hibernate-core/src/main/java/org/hibernate/boot/model/naming/ImplicitNamingStrategyJpaCompliantImpl.java

```diff
@@ -198,7 +198,7 @@ public Identifier determineListIndexColumnName(ImplicitIndexColumnNameSource sou
 
 	@Override
 	public Identifier determineForeignKeyName(ImplicitForeignKeyNameSource source) {
-		return toIdentifier(
+		return source.getUserProvidedIdentifier() != null ? source.getUserProvidedIdentifier() : toIdentifier(
```

Comment:
Assign the return of `getUserProvidedIdentifier()` to a local variable, otherwise it has to be built twice.